### PR TITLE
Delete storage feature

### PR DESCRIPTION
**Pull request by Nagendracv**
Delete storage feature (Draft)

## Description
Added ability to delete a storage blob. The mechanism used is to delete by deleting the tf file and committing the changes.

## How Has This Been Tested?
Yes, tested until the delete succeeds in azure. The Output handler queue has not been tested.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Unit test (new or update to tests)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (change that would cause documentation to be updated)
- [ ] Configuration change (change that would affect the configuration of the application)

## Checklist:
- [x] PR is submitted to the correct branch `develop`.
- [x] Code follows the code style of this project.
- [ ] Changes are covered by Specflow tests and all new or existing tests are passing.
- [ ] Any new or updated translatable strings have been added to the localization files.
- [x] Necessary and relevant documentation has been created or updated.
- [x] There are either no changes to the application's configuration or the necessary changes in the infrastructure repository are included in a separate PR.

### CODE REVIEW

File: Portal/src/Datahub.Functions/TerraformOutputHandler.cs [hunk 1]
```diff
@@ -393,4 +402,39 @@ private static string GetStatusMapping(string value)
...
+            _logger.LogInformation("Project resource with service type {TerraformServiceType} deleted successfully for project acronym {ProjectAcronymValue}",
+                                    terraformServiceType, projectAcronym.Value);
+            return true;
+        }
+        catch (Exception e) {
+            _logger.LogError(e, "Error deleting project resource");
+        }
+
+        return true;
+    }
```
Comment (yjmrobert) [inline Portal/src/Datahub.Functions/TerraformOutputHandler.cs]: Let's do a soft delete like we discussed, where the row isn't removed but the state is marked as deleted

File: Portal/src/Datahub.Infrastructure/Services/RequestManagementService.cs [hunk 2]
```diff
@@ -125,6 +126,13 @@ await ctx.Entry(project)
             var workspaceDefinition =
                 await _resourceMessagingService.GetWorkspaceDefinition(project.Project_Acronym_CD,
                     requestingUser.Email);
+
+            if (isDelete) {
+                foreach (var template in allTemplates) {
+                    if (template.Name == terraformTemplate) template.Status = "delete";
+                }
+            }
+
```
Comment (yjmrobert) [inline Portal/src/Datahub.Infrastructure/Services/RequestManagementService.cs:145]: The status here won't be persisted to the database. It would make more sense to process this up near L108 so that the `ProcessRequest` call can persist the template status into the database

File: Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor [hunk 3]
```diff
@@ -64,7 +63,11 @@
                                 case ToolStatus.Exists:
                                     <MudButton Variant="@Variant.Filled" Color="@Color.Primary" Disabled>
                                         @Localizer["Tool has been created"]
-                                        <MudIcon Icon="fa-light fa-check-to-slot" Class="ml-2" Style="font-size: 0.8rem;"/>
+                                        <MudIcon Icon="fa-light fa-check-to-slot" Class="ml-1" Style="font-size: 0.8rem;"/>
```
Comment (yjmrobert) [inline Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor]: I think a delete style icon might be more appropriate. We have the font-awesome pro library so there are a lot of options

File: Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor [hunk 4]
```diff
@@ -120,7 +123,9 @@
     {
         Available,
         SendingRequest,
+        SendingDelete,
         PendingRequest,
+        PendingDeleteRequest,
```
Comment (yjmrobert) [inline Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor:142]: Are there UI updates for these statuses? I can see the updates but no corresponding blazor

File: ResourceProvisioner/src/ResourceProvisioner.Infrastructure/Services/TerraformService.cs [hunk 5]
```diff
@@ -327,4 +361,43 @@ private string ComputeBackendConfigValue(string workspaceName, string variableNa
...
+                CreateNoWindow = true,
+            }
+        };
+
+        process.Start();
+        await process.WaitForExitAsync();
+
+        return process.ExitCode == 0;
+    }
+
```
Comment (yjmrobert) [inline ResourceProvisioner/src/ResourceProvisioner.Infrastructure/Services/TerraformService.cs]: Comment out

**Review by yjmrobert — CHANGES_REQUESTED** [review]

File: Portal/src/Datahub.Infrastructure/Services/RequestManagementService.cs (hunk 2 above)
Comment (Nagendracv) [reply Portal/src/Datahub.Infrastructure/Services/RequestManagementService.cs:145]: now persisted

File: ResourceProvisioner/src/ResourceProvisioner.Infrastructure/Services/TerraformService.cs (hunk 5 above)
Comment (Nagendracv) [reply ResourceProvisioner/src/ResourceProvisioner.Infrastructure/Services/TerraformService.cs]: commented now

File: Portal/src/Datahub.Functions/TerraformOutputHandler.cs (hunk 1 above)
Comment (Nagendracv) [reply Portal/src/Datahub.Functions/TerraformOutputHandler.cs]: soft delete implemented now

File: Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor (hunk 4 above)
Comment (Nagendracv) [reply Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor:142]: Updated razor code as well

File: Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor (hunk 3 above)
Comment (Nagendracv) [reply Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor]: used Trash Icon

**Comment by yjmrobert** [comment]
@Nagendracv, just to confirm before I review. This feature is complete and were I to run it I would be able to provision a storage for my workspace and delete the storage?

**Comment by Nagendracv** [comment]
> @Nagendracv, just to confirm before I review. This feature is complete and were I to run it I would be able to provision a storage for my workspace and delete the storage?

Yes full cycle is now tested - including the datahub.functions - terraform output handler.
Earlier I had tested with local data using the sudo page expected output. 
Now connected with the azure deployed dev application's output storage queue.
No changes were needed there as well

File: Datahub.sln [hunk 6]
```diff
@@ -127,9 +130,19 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Datahub.Infrastructure.Offl
...
+	ProjectSection(SolutionItems) = preProject
+		ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\at3_workspace_data.json = ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\at3_workspace_data.json
+		ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\configure_env.ps1 = ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\configure_env.ps1
+		ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\function_app_test.py = ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\function_app_test.py
+		ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\README.md = ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\README.md
+		ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\requirements.txt = ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\requirements.txt
+		ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\workspace_data.json = ResourceProvisioner\test\ResourceProvisioner_PyFunctions_Tests\workspace_data.json
+	EndProjectSection
 EndProject
```
Comment (yjmrobert) [inline Datahub.sln]: Not part of this PR. I would recommend using vscode when working with python. No need to add the files to the csharp solution items

File: Portal/src/Datahub.Core/Utils/TerraformOutputHelper.cs
Comment (yjmrobert) [inline Portal/src/Datahub.Core/Utils/TerraformOutputHelper.cs:1]: Please refactor all `(r.Status == null || r.Status != TerraformOutputStatus.Deleted))` into a property method

File: Portal/src/Datahub.Functions/TerraformOutputHandler.cs [hunk 7]
```diff
@@ -236,17 +236,24 @@ internal async Task ProcessAzureStorageBlob(IReadOnlyDictionary<string, Terrafor
...
+            if (isDeleted)
+                _logger.LogInformation("Azure storage blob status is Deleted. Status: {Status}", storageBlobStatus);
+            else
+                _logger.LogError("Azure storage blob deleted from azure but could not be removed from project. Status: {Status}", storageBlobStatus);
             return;
         }
 
-        var projectResource = await GetProjectResource(outputVariables,
-            TerraformTemplate.GetTerraformServiceType(TerraformTemplate.AzureStorageBlob));
+        var projectResource = await GetProjectResource(outputVariables, storageServiceType);
```
Comment (yjmrobert) [inline Portal/src/Datahub.Functions/TerraformOutputHandler.cs]: Please simplify the nesting

File: Portal/src/Datahub.Functions/TerraformOutputHandler.cs [hunk 8]
```diff
@@ -391,6 +400,44 @@ private static string GetStatusMapping(string value)
...
+            _logger.LogInformation("Project resource with service type {TerraformServiceType} deleted successfully for project acronym {ProjectAcronymValue}",
+                                    terraformServiceType, projectAcronym.Value);
+            return true;
+        }
+        catch (Exception e) {
+            _logger.LogError(e, "Error deleting project resource");
+        }
+
+        return true;
+    }
```
Comment (yjmrobert) [inline Portal/src/Datahub.Functions/TerraformOutputHandler.cs]: In our previous call I suggested just putting L428 `projectResource.Status = storageBlobStatus` up on L277 instead of doing all this a second time inside a method, and fixing all the complex nesting.

Why was this not implemented?

**Review by yjmrobert — CHANGES_REQUESTED** [review]
I stopped reviewing after `TerraformOutputHandler.cs`. I've provided feedback on that flow multiple times and it hasn't been implemented. 

Please revisit the feedback given and update

File: Portal/src/Datahub.Functions/TerraformOutputHandler.cs (hunk 8 above)
Comment (Nagendracv) [reply Portal/src/Datahub.Functions/TerraformOutputHandler.cs]: Apologies, for some reason, this particular comment was not visible in my review list. I've updated it, will push in a minute

Comment (yjmrobert) [reply Portal/src/Datahub.Functions/TerraformOutputHandler.cs]: That's not what I said, your update never sets the status unless it is deleted.

image

it should be something like

image

Comment (Nagendracv) [reply Portal/src/Datahub.Functions/TerraformOutputHandler.cs]: No... why? Actually Status field would be used only in case of deleted... well mostly... unless you want to set the status field irrespective of deleted...

your code above is also correct - except that status field will always get set unless its either completed or deleted...
hmm... that amounts to the same thing too... well...

If we are going that route anyways, my thought is, probably, we can make it even more simple
i.e. always set the status field and let the output value from the terraform feed the status field

i.e. this would not be needed... coz it does not mean it has failed. It just means that the output status was set to some other value other than completed and deleted. So I believe we should not 'return' at this juncture, and let just set the status value be set. I think the value will be useful for later troubleshooting as well
image

What are your thoughts on that?

Comment (Nagendracv) [reply Portal/src/Datahub.Functions/TerraformOutputHandler.cs]: like this 
image

Comment (Nagendracv) [reply Portal/src/Datahub.Functions/TerraformOutputHandler.cs]: In that case, it will set all values - including the deleted value. And I guess that's what we also want - especially to handle the failed scenario -  assuming there is another status called Failed status being set in the output value

File: Datahub.sln (hunk 6 above)
Comment (Nagendracv) [reply Datahub.sln]: Done. I had added it here earlier for some reason. But sure it is not needed in the solution.

**Comment by Nagendracv** [comment]
Closing the PR as it was meant only for the task for delete resource. Will do new PR once the whole User Story is complete
